File: archive/cdktf-ts/Pr824/lib/compute.ts
import { AutoscalingGroup } from '@cdktf/provider-aws/lib/autoscaling-group';
import { AutoscalingPolicy } from '@cdktf/provider-aws/lib/autoscaling-policy';
import { DataAwsSsmParameter } from '@cdktf/provider-aws/lib/data-aws-ssm-parameter';
import { IamInstanceProfile } from '@cdktf/provider-aws/lib/iam-instance-profile';
import { IamRole } from '@cdktf/provider-aws/lib/iam-role';
import { IamRolePolicy } from '@cdktf/provider-aws/lib/iam-role-policy';
import { LaunchTemplate } from '@cdktf/provider-aws/lib/launch-template';
import { Lb } from '@cdktf/provider-aws/lib/lb';
import { LbListener } from '@cdktf/provider-aws/lib/lb-listener';
import { LbTargetGroup } from '@cdktf/provider-aws/lib/lb-target-group';
import { AwsProvider } from '@cdktf/provider-aws/lib/provider';
import { Fn, TerraformOutput } from 'cdktf';
import { Construct } from 'constructs';
import { name } from './utils/naming';

export interface ComputeProps {
  provider: AwsProvider;
  environment: string;
  region: string;
  vpcId: string;
  publicSubnets: string[];
  privateSubnets: string[];
  albSgId: string;
  appSgId: string;
  instanceType?: string;
  desiredCapacity?: number;
  minSize?: number;
  maxSize?: number;
  acmCertArn?: string;
}

export class Compute extends Construct {
  public readonly albDns: string;
  public readonly albZoneId: string; // Add this property
  public readonly asgName: string;
  public readonly tgArn: string;

  // Newly added props for Monitoring integration
  public readonly scaleUpPolicyArn: string;
  public readonly scaleDownPolicyArn: string;
  public readonly albTargetGroupName: string;

  constructor(scope: Construct, id: string, props: ComputeProps) {
    super(scope, id);

    const env = props.environment;
    const region = props.region;

    // IAM Role for EC2 with SSM + CW Agent
    const ec2Role = new IamRole(this, 'ec2Role', {
      name: name(env, 'ec2-role', region),
      assumeRolePolicy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Action: 'sts:AssumeRole',
            Effect: 'Allow',
            Principal: { Service: 'ec2.amazonaws.com' },
          },
        ],
      }),
      provider: props.provider,
    });

    new IamRolePolicy(this, 'ec2Policy', {
      name: name(env, 'ec2-cw-ssm-policy', region),
      role: ec2Role.id,
      policy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Effect: 'Allow',
            Action: [
              'ssm:DescribeAssociation',
              'ssm:GetDeployablePatchSnapshotForInstance',
              'ssm:GetDocument',
              'ssm:DescribeDocument',
              'ssm:GetParameters',
              'ssm:ListAssociations',
              'ssm:ListInstanceAssociations',
              'ssm:UpdateInstanceInformation',
              'ec2messages:AcknowledgeMessage',
              'ec2messages:DeleteMessage',
              'ec2messages:FailMessage',
              'ec2messages:GetEndpoint',
              'ec2messages:GetMessages',
              'ec2messages:SendReply',
              'cloudwatch:PutMetricData',
              'logs:CreateLogGroup',
              'logs:CreateLogStream',
              'logs:PutLogEvents',
            ],
            Resource: '*',
          },
        ],
      }),
      provider: props.provider,
    });

    const ec2Profile = new IamInstanceProfile(this, 'ec2Profile', {
      name: name(env, 'ec2-profile', region),
      role: ec2Role.name,
      provider: props.provider,
    });

    // SSM Parameter for AL2023 AMI
    const ssmAmi = new DataAwsSsmParameter(this, 'amiParam', {
      name: '/aws/service/ami-amazon-linux-latest/al2023-ami-kernel-default-x86_64',
      provider: props.provider,
    });

    // ALB
    const alb = new Lb(this, 'alb', {
      name: name(env, 'alb', region),
      internal: false,
      loadBalancerType: 'application',
      securityGroups: [props.albSgId],
      subnets: props.publicSubnets,
      provider: props.provider,
    });

    // Save the DNS and ZoneId
    this.albDns = alb.dnsName;
    this.albZoneId = alb.zoneId; // Ensure the ALB Zone ID is also saved

    // Target Group
    const tg = new LbTargetGroup(this, 'tg', {
      name: name(env, 'tg', region),
      port: 80,
      protocol: 'HTTP',
      vpcId: props.vpcId,
      healthCheck: {
        path: '/',
        protocol: 'HTTP',
      },
      provider: props.provider,
    });

    // Save target group name for monitoring
    this.albTargetGroupName = tg.name;

    // Create HTTPS listener only if a cert was provided
    if (props.acmCertArn && props.acmCertArn.trim().length > 0) {
      new LbListener(this, 'httpsListener', {
        loadBalancerArn: alb.arn,
        port: 443,
        protocol: 'HTTPS',
        certificateArn: props.acmCertArn,
        defaultAction: [
          {
            type: 'forward',
            targetGroupArn: tg.arn,
          },
        ],
      });
    }

    // Launch Template with CW Agent + SSM Agent
    const userData = Fn.base64encode(`#!/bin/bash
yum update -y
yum install -y amazon-cloudwatch-agent
systemctl enable amazon-cloudwatch-agent
systemctl start amazon-cloudwatch-agent
`);

    const lt = new LaunchTemplate(this, 'lt', {
      namePrefix: name(env, 'lt', region),
      imageId: ssmAmi.value,
      instanceType: props.instanceType || 't3.micro',
      vpcSecurityGroupIds: [props.appSgId],
      iamInstanceProfile: { name: ec2Profile.name },
      userData,
      provider: props.provider,
    });

    // ASG
    const asg = new AutoscalingGroup(this, 'asg', {
      name: name(env, 'asg', region),
      minSize: props.minSize ?? 1,
      maxSize: props.maxSize ?? 3,
      desiredCapacity: props.desiredCapacity ?? 1,
      vpcZoneIdentifier: props.privateSubnets,
      launchTemplate: { id: lt.id, version: '$Latest' },
      targetGroupArns: [tg.arn],
      provider: props.provider,
    });

    // Scale policies
    const scaleUpPolicy = new AutoscalingPolicy(this, 'scaleUp', {
      name: name(env, 'scale-up', region),
      scalingAdjustment: 1,
      adjustmentType: 'ChangeInCapacity',
      cooldown: 300,
      autoscalingGroupName: asg.name,
      provider: props.provider,
    });

    const scaleDownPolicy = new AutoscalingPolicy(this, 'scaleDown', {
      name: name(env, 'scale-down', region),
      scalingAdjustment: -1,
      adjustmentType: 'ChangeInCapacity',
      cooldown: 300,
      autoscalingGroupName: asg.name,
      provider: props.provider,
    });

    // Save ARNs for monitoring
    this.scaleUpPolicyArn = scaleUpPolicy.arn;
    this.scaleDownPolicyArn = scaleDownPolicy.arn;

    // Expose DNS, ASG name, TG ARN
    new TerraformOutput(this, 'alb_dns', { value: this.albDns });
    new TerraformOutput(this, 'alb_zone_id', { value: this.albZoneId }); // Add output for ALB Zone ID
    new TerraformOutput(this, 'asg_name', { value: asg.name });
  }
}

File: archive/cdktf-py/Pr7958/lib/compute.py
from constructs import Construct
from cdktf import Fn
from cdktf_cdktf_provider_aws.iam_role import IamRole
from cdktf_cdktf_provider_aws.iam_role_policy_attachment import IamRolePolicyAttachment
from cdktf_cdktf_provider_aws.iam_instance_profile import IamInstanceProfile
from cdktf_cdktf_provider_aws.launch_template import LaunchTemplate, LaunchTemplateIamInstanceProfile
from cdktf_cdktf_provider_aws.lb import Lb
from cdktf_cdktf_provider_aws.lb_target_group import LbTargetGroup, LbTargetGroupHealthCheck
from cdktf_cdktf_provider_aws.lb_listener import LbListener, LbListenerDefaultAction
from cdktf_cdktf_provider_aws.autoscaling_group import AutoscalingGroup, AutoscalingGroupTag
from cdktf_cdktf_provider_aws.data_aws_ami import DataAwsAmi, DataAwsAmiFilter


class ComputeConstruct(Construct):
    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        environment_suffix: str,
        primary_provider,
        secondary_provider,
        primary_public_subnet_ids: list,
        secondary_public_subnet_ids: list,
        primary_private_subnet_ids: list,
        secondary_private_subnet_ids: list,
        primary_db_endpoint: str,
        secondary_db_endpoint: str,
        primary_app_security_group_id: str,
        primary_alb_security_group_id: str,
        secondary_app_security_group_id: str,
        secondary_alb_security_group_id: str,
        primary_vpc_id: str,
        secondary_vpc_id: str
    ):
        super().__init__(scope, construct_id)

        # Lookup latest Amazon Linux 2023 AMI for primary region
        primary_ami = DataAwsAmi(
            self,
            "primary-ami",
            most_recent=True,
            owners=["amazon"],
            filter=[
                DataAwsAmiFilter(
                    name="name",
                    values=["al2023-ami-2023.*-x86_64"]
                ),
                DataAwsAmiFilter(
                    name="virtualization-type",
                    values=["hvm"]
                ),
                DataAwsAmiFilter(
                    name="architecture",
                    values=["x86_64"]
                )
            ],
            provider=primary_provider
        )

        # Lookup latest Amazon Linux 2023 AMI for secondary region
        secondary_ami = DataAwsAmi(
            self,
            "secondary-ami",
            most_recent=True,
            owners=["amazon"],
            filter=[
                DataAwsAmiFilter(
                    name="name",
                    values=["al2023-ami-2023.*-x86_64"]
                ),
                DataAwsAmiFilter(
                    name="virtualization-type",
                    values=["hvm"]
                ),
                DataAwsAmiFilter(
                    name="architecture",
                    values=["x86_64"]
                )
            ],
            provider=secondary_provider
        )

        # IAM role for EC2 instances - Primary
        primary_instance_role = IamRole(
            self,
            "primary-instance-role",
            name=f"primary-instance-role-{environment_suffix}",
            assume_role_policy=Fn.jsonencode({
                "Version": "2012-10-17",
                "Statement": [{
                    "Effect": "Allow",
                    "Principal": {"Service": "ec2.amazonaws.com"},
                    "Action": "sts:AssumeRole"
                }]
            }),
            tags={"Name": f"primary-instance-role-{environment_suffix}"},
            provider=primary_provider
        )

        IamRolePolicyAttachment(
            self,
            "primary-ssm-policy",
            role=primary_instance_role.name,
            policy_arn="arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore",
            provider=primary_provider
        )

        IamRolePolicyAttachment(
            self,
            "primary-cloudwatch-policy",
            role=primary_instance_role.name,
            policy_arn="arn:aws:iam::aws:policy/CloudWatchAgentServerPolicy",
            provider=primary_provider
        )

        primary_instance_profile = IamInstanceProfile(
            self,
            "primary-instance-profile",
            name=f"primary-instance-profile-{environment_suffix}",
            role=primary_instance_role.name,
            provider=primary_provider
        )

        # User data for primary instances
        primary_user_data = Fn.base64encode(f"""#!/bin/bash
yum update -y
yum install -y docker
systemctl start docker
systemctl enable docker
docker run -d -p 8080:8080 \
  -e DB_ENDPOINT={primary_db_endpoint} \
  -e REGION=us-east-1 \
  --name trading-app \
  nginx:latest
""")

        # Launch template - Primary
        primary_launch_template = LaunchTemplate(
            self,
            "primary-launch-template",
            name=f"primary-lt-{environment_suffix}",
            image_id=primary_ami.id,
            instance_type="t3.medium",
            user_data=primary_user_data,
            iam_instance_profile=LaunchTemplateIamInstanceProfile(
                arn=primary_instance_profile.arn
            ),
            vpc_security_group_ids=[primary_app_security_group_id],
            tags={"Name": f"primary-lt-{environment_suffix}"},
            provider=primary_provider
        )

        # ALB - Primary
        self.primary_alb = Lb(
            self,
            "primary-alb",
            name=f"primary-alb-{environment_suffix}",
            internal=False,
            load_balancer_type="application",
            security_groups=[primary_alb_security_group_id],
            subnets=primary_public_subnet_ids,
            enable_deletion_protection=False,
            tags={"Name": f"primary-alb-{environment_suffix}"},
            provider=primary_provider
        )

        # Target group - Primary
        primary_target_group = LbTargetGroup(
            self,
            "primary-tg",
            name=f"primary-tg-{environment_suffix}",
            port=8080,
            protocol="HTTP",
            vpc_id=primary_vpc_id,
            health_check=LbTargetGroupHealthCheck(
                enabled=True,
                path="/health",
                protocol="HTTP",
                interval=30,
                timeout=5,
                healthy_threshold=2,
                unhealthy_threshold=2
            ),
            tags={"Name": f"primary-tg-{environment_suffix}"},
            provider=primary_provider
        )

        # Listener - Primary
        LbListener(
            self,
            "primary-listener",
            load_balancer_arn=self.primary_alb.arn,
            port=80,
            protocol="HTTP",
            default_action=[LbListenerDefaultAction(
                type="forward",
                target_group_arn=primary_target_group.arn
            )],
            provider=primary_provider
        )

        # Auto Scaling Group - Primary
        self.primary_asg = AutoscalingGroup(
            self,
            "primary-asg",
            name=f"primary-asg-{environment_suffix}",
            vpc_zone_identifier=primary_private_subnet_ids,
            min_size=2,
            max_size=6,
            desired_capacity=2,
            health_check_type="ELB",
            health_check_grace_period=300,
            target_group_arns=[primary_target_group.arn],
            launch_template={"id": primary_launch_template.id, "version": "$Latest"},
            tag=[
                AutoscalingGroupTag(
                    key="Name",
                    value=f"primary-asg-{environment_suffix}",
                    propagate_at_launch=True
                )
            ],
            provider=primary_provider
        )

        # IAM role for EC2 instances - Secondary
        secondary_instance_role = IamRole(
            self,
            "secondary-instance-role",
            name=f"secondary-instance-role-{environment_suffix}",
            assume_role_policy=Fn.jsonencode({
                "Version": "2012-10-17",
                "Statement": [{
                    "Effect": "Allow",
                    "Principal": {"Service": "ec2.amazonaws.com"},
                    "Action": "sts:AssumeRole"
                }]
            }),
            tags={"Name": f"secondary-instance-role-{environment_suffix}"},
            provider=secondary_provider
        )

        IamRolePolicyAttachment(
            self,
            "secondary-ssm-policy",
            role=secondary_instance_role.name,
            policy_arn="arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore",
            provider=secondary_provider
        )

        IamRolePolicyAttachment(
            self,
            "secondary-cloudwatch-policy",
            role=secondary_instance_role.name,
            policy_arn="arn:aws:iam::aws:policy/CloudWatchAgentServerPolicy",
            provider=secondary_provider
        )

        secondary_instance_profile = IamInstanceProfile(
            self,
            "secondary-instance-profile",
            name=f"secondary-instance-profile-{environment_suffix}",
            role=secondary_instance_role.name,
            provider=secondary_provider
        )

        # User data for secondary instances
        secondary_user_data = Fn.base64encode(f"""#!/bin/bash
yum update -y
yum install -y docker
systemctl start docker
systemctl enable docker
docker run -d -p 8080:8080 \
  -e DB_ENDPOINT={secondary_db_endpoint} \
  -e REGION=us-east-2 \
  --name trading-app \
  nginx:latest
""")

        # Launch template - Secondary
        secondary_launch_template = LaunchTemplate(
            self,
            "secondary-launch-template",
            name=f"secondary-lt-{environment_suffix}",
            image_id=secondary_ami.id,
            instance_type="t3.medium",
            user_data=secondary_user_data,
            iam_instance_profile=LaunchTemplateIamInstanceProfile(
                arn=secondary_instance_profile.arn
            ),
            vpc_security_group_ids=[secondary_app_security_group_id],
            tags={"Name": f"secondary-lt-{environment_suffix}"},
            provider=secondary_provider
        )

        # ALB - Secondary
        self.secondary_alb = Lb(
            self,
            "secondary-alb",
            name=f"secondary-alb-{environment_suffix}",
            internal=False,
            load_balancer_type="application",
            security_groups=[secondary_alb_security_group_id],
            subnets=secondary_public_subnet_ids,
            enable_deletion_protection=False,
            tags={"Name": f"secondary-alb-{environment_suffix}"},
            provider=secondary_provider
        )

        # Target group - Secondary
        secondary_target_group = LbTargetGroup(
            self,
            "secondary-tg",
            name=f"secondary-tg-{environment_suffix}",
            port=8080,
            protocol="HTTP",
            vpc_id=secondary_vpc_id,
            health_check=LbTargetGroupHealthCheck(
                enabled=True,
                path="/health",
                protocol="HTTP",
                interval=30,
                timeout=5,
                healthy_threshold=2,
                unhealthy_threshold=2
            ),
            tags={"Name": f"secondary-tg-{environment_suffix}"},
            provider=secondary_provider
        )

        # Listener - Secondary
        LbListener(
            self,
            "secondary-listener",
            load_balancer_arn=self.secondary_alb.arn,
            port=80,
            protocol="HTTP",
            default_action=[LbListenerDefaultAction(
                type="forward",
                target_group_arn=secondary_target_group.arn
            )],
            provider=secondary_provider
        )

        # Auto Scaling Group - Secondary
        self.secondary_asg = AutoscalingGroup(
            self,
            "secondary-asg",
            name=f"secondary-asg-{environment_suffix}",
            vpc_zone_identifier=secondary_private_subnet_ids,
            min_size=2,
            max_size=6,
            desired_capacity=2,
            health_check_type="ELB",
            health_check_grace_period=300,
            target_group_arns=[secondary_target_group.arn],
            launch_template={"id": secondary_launch_template.id, "version": "$Latest"},
            tag=[
                AutoscalingGroupTag(
                    key="Name",
                    value=f"secondary-asg-{environment_suffix}",
                    propagate_at_launch=True
                )
            ],
            provider=secondary_provider
        )

    @property
    def primary_alb_dns(self):
        return self.primary_alb.dns_name

    @property
    def secondary_alb_dns(self):
        return self.secondary_alb.dns_name

    @property
    def primary_alb_zone_id(self):
        return self.primary_alb.zone_id

    @property
    def secondary_alb_zone_id(self):
        return self.secondary_alb.zone_id

    @property
    def primary_alb_arn_suffix(self):
        return self.primary_alb.arn_suffix

    @property
    def secondary_alb_arn_suffix(self):
        return self.secondary_alb.arn_suffix

    @property
    def primary_alb_full_arn(self):
        return self.primary_alb.arn

    @property
    def secondary_alb_full_arn(self):
        return self.secondary_alb.arn

    @property
    def primary_asg_name(self):
        return self.primary_asg.name

    @property
    def secondary_asg_name(self):
        return self.secondary_asg.name
